python/cugraph/structure/number_map.py

```diff
@@ -0,0 +1,716 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cudf
+import dask_cudf
+import numpy as np
+import bisect
+
+
+class NumberMap:
+    class SingleGPU:
+        def __init__(self, df, src_col_names, dst_col_names):
```

Comment:
Can we add an additional parameter to sort IDs based on in-degree or out-degree?

If we want to insure that higher degree vertices have a smaller ID after renumbering (within a segment if we partition vertices to multiple segments especially for multi-GPU case), I guess we can do something like ```df[src_col_names].groupby(src_col_names).size()``` and sort the resulting DataFrame by size (which is degree).

Things are actually more complicated as we need to consider df[dst_col_names] as well... but I think there is a way to handle this... If we can do this with relatively small performance overhead, this can be used to reduce memory footprint for our Graph class and this will make the pattern accelerator optimization simpler (in code complexity) and faster with smaller memory footprint.

If we intend to generate a CSR-ish graph (to implement the push model, store the graph adjacency matrix in row-major order), we need to sort based on out-degree, and if we intend to generate a CSC-ish graph (to implement the pull model, store the graph adjacency matrix in column-major order), we need to sort based on in-degree.